From: Flavio Leitner <fbl@redhat.com>
Date: Tue, 5 Mar 2013 08:11:01 +0000
Subject: tcp: ipv6: bind() use stronger condition for bind_conflict
Patch-mainline: v3.10-rc1
Git-commit: dd9f319d94c99b96fc9b34ccde7389a91059fe31
References: bnc#823618

We must try harder to get unique (addr, port) pairs when
doing port autoselection for sockets with SO_REUSEADDR
option set.

This is a continuation of commit aacd9289af8b82f5fb01bcdd53d0e3406d1333c7
for IPv6.

Signed-off-by: Flavio Leitner <fbl@redhat.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
Signed-off-by: Ales Novak <alnovak@suse.cz>
Signed-off-by: Michal Kubecek <mkubecek@suse.cz>

Conflicts:
	net/ipv6/inet6_connection_sock.c
---
 net/ipv6/inet6_connection_sock.c | 15 ++++++++++-----
 1 file changed, 10 insertions(+), 5 deletions(-)

diff --git a/net/ipv6/inet6_connection_sock.c b/net/ipv6/inet6_connection_sock.c
index 2c2f643..ab64d50 100644
--- a/net/ipv6/inet6_connection_sock.c
+++ b/net/ipv6/inet6_connection_sock.c
@@ -43,11 +43,16 @@ int inet6_csk_bind_conflict(const struct sock *sk,
 		if (sk != sk2 &&
 		    (!sk->sk_bound_dev_if ||
 		     !sk2->sk_bound_dev_if ||
-		     sk->sk_bound_dev_if == sk2->sk_bound_dev_if) &&
-		    (!sk->sk_reuse || !sk2->sk_reuse ||
-		     sk2->sk_state == TCP_LISTEN) &&
-		     ipv6_rcv_saddr_equal(sk, sk2))
-			break;
+		     sk->sk_bound_dev_if == sk2->sk_bound_dev_if)) {
+			if ((!sk->sk_reuse || !sk2->sk_reuse ||
+			     sk2->sk_state == TCP_LISTEN) &&
+			      ipv6_rcv_saddr_equal(sk, sk2))
+				break;
+			if (!relax && sk->sk_reuse && sk2->sk_reuse &&
+			    sk2->sk_state != TCP_LISTEN &&
+			    ipv6_rcv_saddr_equal(sk, sk2))
+				break;
+		}
 	}
 
 	return node != NULL;
-- 
1.8.1.4

